File: ga/glm4_result_2024_6/result/multi_thread_basic_cpp_glm/13_multi_thread_basic.cpp
use std::sync::{Arc, atomic::AtomicI32};
use std::thread;

fn withdraw(account: Arc<AtomicI32>, amount: i32) {
    for _ in 0..amount {
        account.fetch_sub(1, std::sync::atomic::Ordering::SeqCst);
    }
}

fn main() {
    let account_balance = Arc::new(AtomicI32::new(1000000));

    let account1 = Arc::clone(&account_balance);
    let handle1 = thread::spawn(move || {
        withdraw(account1, 500000);
    });

    let account2 = Arc::clone(&account_balance);
    let handle2 = thread::spawn(move || {
        withdraw(account2, 500000);
    });

    handle1.join().unwrap();
    handle2.join().unwrap();

    println!("Final account balance: {}", account_balance.load(std::sync::atomic::Ordering::SeqCst));
}

File: ga/4omini_result_2024_7/result/multi_thread_basic_mini/13_multi_thread_basic.rs
use std::sync::{Arc, Mutex};
use std::thread;

fn withdraw(account_balance: Arc<Mutex<i32>>, amount: i32) {
    for _ in 0..amount {
        let mut balance = account_balance.lock().unwrap();
        *balance -= 1;
    }
}

fn main() {
    let account_balance = Arc::new(Mutex::new(1_000_000)); // Initial account balance

    let amount_to_withdraw = 500_000;

    let account_balance_clone1 = Arc::clone(&account_balance);
    let thread1 = thread::spawn(move || {
        withdraw(account_balance_clone1, amount_to_withdraw);
    });

    let account_balance_clone2 = Arc::clone(&account_balance);
    let thread2 = thread::spawn(move || {
        withdraw(account_balance_clone2, amount_to_withdraw);
    });

    thread1.join().unwrap();
    thread2.join().unwrap();

    // Display the final account balance
    let final_balance = account_balance.lock().unwrap();
    println!("Final account balance: {}", *final_balance);
}